bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+
+''' P2TR (key path spends)'''
+
+The sender MUST use the tweaked private key corresponding to the taproot output key (i.e. the tweaked private key for a key path spend). If the taproot output key is not available, the output cannot be used for sending a silent payment. The receiver always uses the taproot output key when scanning, regardless of whether the taproot output is being used in a key path spend or a script path spend<ref name="why_always_output_pubkey">''' Why not skip all taproot script path spends? ''' This causes malleability issues for CoinJoins. If the silent payments protocol skipped taproot script path spends, this would allow an attacker to join a CoinJoin round, participate in deriving the silent payment address using the tweaked private key for a key path spend, and then broadcast their own version of the transaction using the script path spend. If the receiver were to only consider key path spends, they would skip the attacker's script path spend input when deriving the shared secret and not be able to find the funds.</ref>.
+
+The one exception is script path spends that use NUMS point ''H'' as their internal key (as defined in [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs BIP341: Constructing and spending Taproot outputs]), in which case the output will be skipped for the purposes of shared secret derivation<ref name="why_ignore_h">'''Why skip outputs with H as the internal taproot key?''' If use cases get popularized where the taproot key path cannot be used, these outputs can still be included without getting in the way of making a silent payment, provided they specifically use H as their internal taproot key.</ref>.
+
+''' P2WPKH '''
+
+The sender performs the tweak using the private key for the output and the receiver obtains the public key from the ''witness''.
+
+''' P2SH-P2WPKH '''
+
+The sender performs the tweak using the private key for the nested ''P2WPKH'' output and the receiver obtains the public key from the ''witness''.
+
+''' P2PKH '''
+
+The sender performs the tweak using the private key for the output and SHOULD sign using the standard script template:
+
+
+    scriptSig: <Signature> <Public Key>
+
+
+The receiver obtains the public key from the ''scriptSig''. The receiver MUST parse the ''scriptSig'' for the public key, even if the ''scriptSig'' is non-standard (e.g. <code><dummy> OP_DROP <Signature> <Public Key></code>). This is to address the [https://en.bitcoin.it/wiki/Transaction_malleability third-party malleability of ''P2PKH'' ''scriptSigs''].
+
+=== Sender ===
+
+==== Selecting inputs ====
+
+The sending wallet performs coin selection as usual with the following restrictions:
+
+* At least one input MUST be from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* Exclude ''P2TR'' script path spends, unless ''H'' is used as the internal public key
```

Comment:
Should this read
> Exclude ''P2TR'' outputs for whom `ax * Bscan` cannot be computed

?

Reply:
We want to be clear that the sender should never use the script path spend when funding a silent payment transaction. This is because a script path spend could contain any number of public keys and it becomes intractable for the receiver to try and determine which ones would have been used.

In the event the sender uses a script path spend where the internal public keys is H, it is safe to include it to fund the transaction because the receiver can clearly skip it when collecting the public keys for the ECDH step.

Reply:
Hmm, can we instead put it on the recipient to exclude P2TR inputs from the tweak if they are spent with script path, but let the sender include them?

Any exclusion of allowable inputs seems like a footgun.

Reply:
That would be ideal! Unfortunately, that causes a malleability issue with CoinJoin. This is explained in one of the footones: 

> Why not skip all taproot script path spends? This causes malleability issues for CoinJoins. If the silent payments protocol skipped taproot script path spends, this would allow an attacker to join a CoinJoin round, participate in deriving the silent payment address using the tweaked private key for a key path spend, and then broadcast their own version of the transaction using the script path spend. If the receiver were to only consider key path spends, they would skip the attacker's script path spend input when deriving the shared secret and not be able to find the funds.

Reply:
>one can imagine a CKBunker-type device which would be willing to provide an answer to the question "What is your key times this public point" but wouldn't be willing to sign a specific transaction spending its inputs. Then the spender could collect whatever other spending conditions might let them spend from the script path and still send the silent payment.

The specification as currently described already allows for this. Was there any part that gave you the impression that it didn't?

Reply:
> The specification as currently described already allows for this. Was there any part that gave you the impression that it didn't?

`Exclude ''P2TR" script path spends` <-- that phrase in the input selection

The requirement is that even if I am going to spend with script path I must include the taproot outputkey in the ECDH to prevent malleability.

Reply:
> a CKBunker-type device which would be willing to provide an answer to the question "What is your key times this public point" but wouldn't be willing to sign a specific transaction spending its inputs. Then the spender could collect whatever other spending conditions might let them spend from the script path and still send the silent payment.

If I'm understanding your example, the key path private key is on the CKBunker and never signs transactions, but the user can spend from a separate wallet using script path conditions for the same input(s)?

Reply:
> If I'm understanding your example, the key path private key is on the CKBunker and never signs transactions, but the user can spend from a separate wallet using script path conditions for the same input(s)?

Yeah. I might have my CKBunker at home, set to sign nothing, but produce the ECDH shares for silent payments. I also have an `OP_CSV <pubkey> OP_CHECKSIG` script path for a key I keep on my phone 1 million sats per week, so I can send using my phone and silent payments up to that amount as long as my CKBunker is willing to provide its ECDH shares.

Obviously I made up this example on the spot, but with vaults and other advanced constructions, I don't think it makes sense to exclude the possibility that someone could have limited access to the keypath secret key (or shares), but want to spend using a script path for one reason or another.

Reply:
Thanks for the explanation! Initially, my thinking on this was there's never a (likely) scenario where you would have access to the taproot output key but choose to spend with the script path, so we could likely simplify implementations for the sender by just skipping inputs where we only have the script path data altogether in coin selection.

But, given your example, I agree that there might be interesting use cases that we don't want to rule out.

Reply:
Perhaps as a foot note to point out there are theoretical exceptions to `Exclude ''P2TR" script path spends`. In practice almost none of the sender implementations are going to need to consider this, but it's true, so worth mentioning.

Reply:
I updated the wording to remove the "exclude" language, more as a placeholder because I think this still needs a bit of re-working (along with the "P2TR" section in "Inputs for Shared Secret Derivation"